### PR TITLE
Add LspUpdateServers

### LINKED ISSUE #315: add command to allow updating all installed language server
I use multiple language servers so would be good to have a single command that I can run and update all installed language server `LspUpdateServers`

### PR DESCRIPTION
**Pull request by mattn**
Fixes #315

**DO NOT USE**

image

### CODE REVIEW

**Comment by prabirshrestha** [comment]
At least on linux doesn't seem to work correctly. This is what I see after using `LspUpdateServers`. 
Haven't tried on other OS yet.

image

**Comment by mattn** [comment]
Sorry, current implementation is wrong.

**Comment by mattn** [comment]
Fixed. But this take very long time to update all servers...

**Comment by prabirshrestha** [comment]
Seems like you are currently using `system`. have you tried using the async version instead?

I recently added support for `lsp#callbag#spawn` which uses job internally and works in neovim and vim. 

```
lsp#callbag#spawn(['bash', '-c', 'echo hello'], { 'exit': 1})
```

https://github.com/prabirshrestha/vim-lsp/pull/951

**Comment by mattn** [comment]
`lsp#callbag#spawn` can do callback?

**Comment by prabirshrestha** [comment]
yes callbag is supported. you need to explictly opt in.


```
function! s:log(...) abort
    echom json_encode(a:000)
endfunction

let s:Stdin = lsp#callbag#makeSubject()

call lsp#callbag#pipe(
    \ lsp#callbag#spawn(['bash', '-c', 'read i; echo $i'], { 'stdin': s:Stdin, 'stdout': 1, 'exit': 1, 'stderr': 1, 'normalize': 'string' }),
    \ lsp#callbag#subscribe({
    \   'next':{x->s:log('next', x)},
    \   'complete':{->s:log('complete')},
    \   'error':{x->s:log('error', x)},
    \ }),
    \ )

call s:Stdin(1, 'hello')
call s:Stdin(2, lsp#callbag#undefined()) " required to close stdin
```

next will get events like `{ 'event': 'stdout', data: '' }` .

Here is the full list of options as shown in the comments of callbag.vim

```
spawn {{{
" let s:Stdin = lsp#callbag#makeSubject()
" call lsp#callbag#spawn(['bash', '-c', 'read i; echo $i'], {
"   \ 'stdin': s:Stdin,
"   \ 'stdout': 0,
"   \ 'stderr': 0,
"   \ 'exit': 0,
"   \ 'start': 0, " when job starts before subscribing to stdin
"   \ 'ready': 0, " when job starts and after subscribing to stdin
"   \ 'pid': 0,
"   \ 'failOnNonZeroExitCode': 1,
"   \ 'failOnStdinError': 1,
"   \ 'normalize': 'raw' | 'string' | 'array', (defaults to raw),
"   \ 'env': {},
"   \ })
"   call s:Stdin(1, 'hi')
"   call s:Stdin(2, lsp#callbag#undefined()) " requried to close stdin
```

File: autoload/lsp_settings.vim [hunk 1]
```diff
@@ -514,6 +514,50 @@ function! lsp_settings#clear() abort
   let s:ftmap = {}
 endfunction
 
+function! lsp_settings#update_servers(bang) abort
+  if empty(a:bang) && confirm('Update all installed servers?', "&Yes\n&Cancel") ==# 2
+    return
+  endif
+  let l:servers = lsp_settings#installed_servers()
```
Comment (rafi) [inline autoload/lsp_settings.vim:521]: @mattn need to deduplicated servers names:
```viml
let l:servers = uniq(sort(map(lsp_settings#installed_servers(), {_, v -> v.command})))
```

**Comment by rafi** [comment]
@mattn Instead of deleting server directory and reinstalling, why not change `npm-install.*` to install package with `@latest` tag version, e.g. `npm install "${2}@latest"`, this could save a lot of time reinstalling all the `node_modules` directories. Also my comment above regarding deducing duplicated server names.

**Comment by mattn** [comment]
I gave up this feature. But added LspManageServers. So closing

https://github.com/mattn/vim-lsp-settings/pull/510

**Pull request closed without merging**